websocketOutportDriver/client/wsClient.go

```diff
@@ -0,0 +1,188 @@
+		wsConn:                   args.WSConnClient,
+		payloadParser:            args.PayloadParser,
+		operationHandler:         args.OperationHandler,
+		uint64ByteSliceConverter: args.Uint64ByteSliceConverter,
+	}, nil
+}
+
+// Start will start the client listening ws process
+func (c *client) Start() {
+	log.Info("connecting to", "url", c.url)
+
+	for {
+		err := c.wsConn.OpenConnection(c.url)
+		if err != nil {
+			log.Warn(fmt.Sprintf("c.openConnection(), retrying in %v...", c.retryDuration), "error", err.Error())
+			time.Sleep(c.retryDuration)
+			continue
+		}
+
+		closed := c.listenOnWebSocket()
+		if closed {
+			return
+		}
+	}
+}
+
+func (c *client) listenOnWebSocket() (closed bool) {
+	for {
+		_, message, err := c.wsConn.ReadMessage()
+		if err == nil {
+			c.verifyPayloadAndSendAckIfNeeded(message)
+			continue
+		}
+
+		_, isConnectionClosed := err.(*websocket.CloseError)
+		if !isConnectionClosed {
+			if strings.Contains(err.Error(), closedConnection) {
+				return true
+			}
+			log.Warn("c.listenOnWebSocket()-> connection problem, retrying", "error", err.Error())
+		} else {
+			log.Warn(fmt.Sprintf("websocket terminated by the server side, retrying in %v...", c.retryDuration), "error", err.Error())
+		}
+		return
+	}
+}
+
+func (c *client) verifyPayloadAndSendAckIfNeeded(payload []byte) {
+	if len(payload) == 0 {
+		log.Error("empty payload")
+		return
+	}
+
+	payloadData, err := c.payloadParser.ExtractPayloadData(payload)
+	if err != nil {
+		log.Error("error while extracting payload data: " + err.Error())
+		return
+	}
+
+	log.Info("processing payload",
+		"counter", payloadData.Counter,
+		"operation type", payloadData.OperationType,
+		"payload", payloadData.Payload,
+		"message length", len(payloadData.Payload),
+	)
+
+	function, ok := c.operationHandler.GetOperationHandler(payloadData.OperationType)
+	if !ok {
+		log.Warn("invalid operation", "operation type", payloadData.OperationType.String())
+		c.sendAckIfNeeded(payloadData, true)
+		return
+	}
+
+	err = function(payloadData.Payload)
+	if err != nil {
+		log.Error("could not process payload", "error", err.Error())
+		c.sendAckIfNeeded(payloadData, true)
+		return
+	}
+
+	c.sendAckIfNeeded(payloadData, false)
+}
+
+func (c *client) sendAckIfNeeded(payloadData *websocketOutportDriver.PayloadData, hadError bool) {
+	if !payloadData.WithAcknowledge {
+		return
+	}
+
+	if hadError && c.blockingOnError {
+		return
+	}
+
+	c.waitForAckSignal(payloadData.Counter)
+}
+
+func (c *client) waitForAckSignal(counter uint64) {
+	for {
+		counterBytes := c.uint64ByteSliceConverter.ToByteSlice(counter)
+		err := c.wsConn.WriteMessage(websocket.BinaryMessage, counterBytes)
+		if err != nil {
+			log.Error("could not write acknowledge message",
+				"error", err.Error(), "retrying in", c.retryDuration)
+
+			time.Sleep(time.Second * c.retryDuration)
+			continue
+		}
+
+		return
+	}
+}
+
+// Close will close the underlying ws connection
+func (c *client) Close() {
```

Comment:
can be done in the next PR: we need a way to make all the calls that started for loops to end when calling this Close function. Solution: reuse the component core/closing/safeChanCloser as a subcomponent for this client. Whenever you need a time.sleep in the for loops, use a time.Timer and a select operation along with a call to ChanClose(), handling the later to do a break on the for loop. This will make all the external calls to this component to finish almost instantly, regardless if the call handled an error operation or not.